### PR TITLE
All changes made for format string component and alignment component

### LINKED ISSUE #2289: SA1011: Space after a bracket for format string component
SA1011 is being trigger for lines that have a closing square bracket and a format string component, like this: `$"{row[ColOther]:C}";`
StyleCop seems to want this: `$"{row[ColOther] :C}";`
However, the auto-formatter in VS2015 removes the extra space between the closing bracket and the colon. I did not see an option in the auto-formatter to add a space in this situation. Is the extra space desirable?

### PR DESCRIPTION
**Pull request by THammond9**
Fixes #2289 

**SA1001**:
Allow 
```csharp
$"{x,3}"
```
Disallow
```csharp
$"{x, 3}"
$"{x ,3}"
```
**SA1011**:
Allow 
```csharp
$"{x[n],3}"
$"{x[n]:N}"
```

**SA1024**:
Disallow 
```csharp
$"{x :N}"
```

Still allowed
```csharp
$"{x: N}"
```

### CODE REVIEW

File: StyleCop.Analyzers/StyleCop.Analyzers.Test/SpacingRules/SA1001UnitTests.cs [hunk 1]
```diff
@@ -164,6 +163,32 @@ public class SA1001UnitTests : CodeFixVerifier
...
+        {
+            string statement = @"var x = new[] { 1, 2, 3, 4, 5, 6, 7 };
+            var t = $""{x[2], 3}"";";
+            string fixedStatement = @"var x = new[] { 1, 2, 3, 4, 5, 6, 7 };
+            var t = $""{x[2],3}"";";
+
+            DiagnosticResult expected = this.CSharpDiagnostic().WithArguments(" not", "followed").WithLocation(8, 28);
+
+            await this.TestCommaInStatementOrDeclAsync(statement, expected, fixedStatement).ConfigureAwait(false);
+        }
```
Comment (vweijsters) [inline StyleCop.Analyzers/StyleCop.Analyzers.Test/SpacingRules/SA1001UnitTests.cs]: Please add a testcase for "{x[2] , 3}" as well to verify that a double fault is detected properly as well.

File: StyleCop.Analyzers/StyleCop.Analyzers.Test/SpacingRules/SA1011UnitTests.cs [hunk 2]
```diff
@@ -312,6 +311,23 @@ void Method()
...
+        public async Task TestColonCanFollowSquareBracketWhenPartOfInterpolationFormatClauseAsync()
+        {
+            string testCode = @"
+class ClassName
+{
+    void Method()
+    {
+        var x = new[] { 1, 2, 3, 4, 5, 6, 7 };
+        var t = $""{ x[2]:C}"";
+    }
```
Comment (vweijsters) [inline StyleCop.Analyzers/StyleCop.Analyzers.Test/SpacingRules/SA1011UnitTests.cs]: Please add a testcase for `$"{x[n],3}"` as well to signify that this is a supported option.

File: StyleCop.Analyzers/StyleCop.Analyzers/SpacingRules/SA1001CommasMustBeSpacedCorrectly.cs [hunk 3]
```diff
@@ -56,12 +57,12 @@ private static void HandleSyntaxTree(SyntaxTreeAnalysisContext context)
...
-                    break;
+                    case SyntaxKind.CommaToken:
+                        HandleCommaToken(context, token);
+                        break;
 
-                default:
-                    break;
+                    default:
+                        break;
                 }
```
Comment (vweijsters) [inline StyleCop.Analyzers/StyleCop.Analyzers/SpacingRules/SA1001CommasMustBeSpacedCorrectly.cs]: Please revert the alignment of the case statements to the original alignment.

File: StyleCop.Analyzers/StyleCop.Analyzers/SpacingRules/SA1001CommasMustBeSpacedCorrectly.cs [hunk 4]
```diff
@@ -16,8 +16,9 @@ namespace StyleCop.Analyzers.SpacingRules
     /// <remarks>
     /// <para>A violation of this rule occurs when the spacing around a comma is incorrect.</para>
     ///
-    /// <para>A comma should always be followed by a single space, unless it is the last character on the line, and a
-    /// comma should never be preceded by any whitespace, unless it is the first character on the line.</para>
+    /// <para>A comma should always be followed by a single space, unless it is the last character on the line
+    /// or it is part of an alignment component, and a comma should never be preceded by any whitespace,
```
Comment (vweijsters) [inline StyleCop.Analyzers/StyleCop.Analyzers/SpacingRules/SA1001CommasMustBeSpacedCorrectly.cs]: I would prefer `string interpolation alignment component` instead of `alignment component` to more clearly state the exact case in which this is allowed.

File: StyleCop.Analyzers/StyleCop.Analyzers/SpacingRules/SA1011ClosingSquareBracketsMustBeSpacedCorrectly.cs [hunk 5]
```diff
@@ -22,8 +21,9 @@ namespace StyleCop.Analyzers.SpacingRules
     /// line.</para>
     ///
     /// <para>A closing square bracket must be followed by whitespace, unless it is the last character on the line, it
-    /// is followed by a closing bracket or an opening parenthesis, it is followed by a comma or semicolon, or it is
-    /// followed by certain types of operator symbols.</para>
+    /// is followed by a closing bracket or an opening parenthesis, it is followed by a comma or semicolon, it is
+    /// followed by a alignment component or format string component, or it is followed by certain types of operator
```
Comment (vweijsters) [inline StyleCop.Analyzers/StyleCop.Analyzers/SpacingRules/SA1011ClosingSquareBracketsMustBeSpacedCorrectly.cs]: I would prefer to see `string interpolation alignment component or string interpolation formatting component`, instead of `alignment component or format string component`.

File: StyleCop.Analyzers/StyleCop.Analyzers/SpacingRules/SA1011ClosingSquareBracketsMustBeSpacedCorrectly.cs [hunk 6]
```diff
@@ -95,37 +95,42 @@ private static void HandleCloseBracketToken(SyntaxTreeAnalysisContext context, S
...
+                        break;
+
+                    case SyntaxKind.ColonToken:
+                        precedesSpecialCharacter = nextToken.Parent.IsKind(SyntaxKind.InterpolationFormatClause);
+                        suppressFollowingSpaceError = false;
+                        break;
+
+                    default:
+                        precedesSpecialCharacter = false;
+                        break;
```
Comment (vweijsters) [inline StyleCop.Analyzers/StyleCop.Analyzers/SpacingRules/SA1011ClosingSquareBracketsMustBeSpacedCorrectly.cs]: Please revert the alignment of the case statements to the original alignment.

File: StyleCop.Analyzers/StyleCop.Analyzers/SpacingRules/SA1024ColonsMustBeSpacedCorrectly.cs [hunk 7]
```diff
@@ -42,6 +41,12 @@ namespace StyleCop.Analyzers.SpacingRules
     /// }
     /// </code>
     ///
+    /// A colon that appears as part of format string component must not have leading whitespace characters. For example:
```
Comment (vweijsters) [inline StyleCop.Analyzers/StyleCop.Analyzers/SpacingRules/SA1024ColonsMustBeSpacedCorrectly.cs]: I would prefer `string interpolation formatting component` instead of `format string component`.

File: StyleCop.Analyzers/StyleCop.Analyzers/SpacingRules/SA1024ColonsMustBeSpacedCorrectly.cs [hunk 8]
```diff
@@ -99,26 +104,32 @@ private static void HandleColonToken(SyntaxTreeAnalysisContext context, SyntaxTo
...
+                    break;
+
+                case SyntaxKind.InterpolationFormatClause:
+                    requireBefore = false;
+                    checkRequireAfter = false;
+                    break;
+
+                default:
+                    return;
             }
```
Comment (vweijsters) [inline StyleCop.Analyzers/StyleCop.Analyzers/SpacingRules/SA1024ColonsMustBeSpacedCorrectly.cs]: Please revert the alignment of the case statements to the original alignment.

File: documentation/SA1001.md [hunk 9]
```diff
@@ -27,6 +27,7 @@ A comma should be followed by a single space, except in the following cases.
 
 * A comma may appear at the end of a line
 * A comma should not be followed by a space when used in an open generic type in a `typeof` expression
+* A comma is part of an alignment component. For example:`$"{x,3}"`
```
Comment (vweijsters) [inline documentation/SA1001.md]: I would prefer `a string interpolation alignment component`, instead of `an alignment component`

File: documentation/SA1011.md [hunk 10]
```diff
@@ -25,7 +25,12 @@ A violation of this rule occurs when the spacing around a closing square bracket
 
 A closing square bracket must never be preceded by whitespace, unless it is the first character on the line.
 
-A closing square bracket must be followed by whitespace, unless it is the last character on the line, it is followed by a closing bracket or an opening parenthesis, it is followed by a comma or semicolon, or it is followed by certain types of operator symbols.
+A closing square bracket must be followed by whitespace unless:
+* It is the last character on the line
+* It is followed by a closing bracket or an opening parenthesis
+* It is followed by a comma or semicolon
+* It is followed by a alignment component or format string component. For example: `$"{x[i],3}"` or `$"{x[i]:C}"`
```
Comment (vweijsters) [inline documentation/SA1011.md]: I would prefer `a string interpolation alignment component or string interpolation formatting component` instead of `a alignment component or format string component`

File: documentation/SA1024.md [hunk 11]
```diff
@@ -45,6 +45,11 @@ switch (x)
 }
 ```
 
+A colon that appears as part of format string component must not have leading whitespace characters. For example:
```
Comment (vweijsters) [inline documentation/SA1024.md]: I would prefer `a string interpolation formatting component` instead of `format string component`

**Comment by sharwell** [comment]
I just got a SA1021 report on [this line](https://github.com/dotnet/roslyn/blob/35afa309845740382633230e7d8cdd6417e9d195/src/Tools/Source/RunTests/TestRunner.cs#L133). Need to add a test case for a negative number after the `,` in an interpolated string.

**Comment by THammond9** [comment]
SA1021 seems to be the correct error for that line. There is no space after the comma and before the negative.

**Comment by sharwell** [comment]
@THammond9 Seems you are correct. A space is allowed after the `,` in that location.

**Comment by THammond9** [comment]
@sharwell Actually the auto format in VS removes the space after the comma so you are still correct that we need a test case for this scenario.

**Comment by THammond9** [comment]
The changes are mostly finished. Just waiting on confirmation before I do anything with them.

File: StyleCop.Analyzers/StyleCop.Analyzers.Test/SpacingRules/SA1001UnitTests.cs (hunk 1 above)
Comment (sharwell) [reply StyleCop.Analyzers/StyleCop.Analyzers.Test/SpacingRules/SA1001UnitTests.cs]: ➡️ Covered in `TestSpaceBeforeAndAfterCommaWhenPartOfInterpolationAlignmentClauseAsync`

File: StyleCop.Analyzers/StyleCop.Analyzers.Test/SpacingRules/SA1011UnitTests.cs (hunk 2 above)
Comment (sharwell) [reply StyleCop.Analyzers/StyleCop.Analyzers.Test/SpacingRules/SA1011UnitTests.cs]: ➡️ Covered in `TestCommaCanFollowSquareBracketWhenPartOfInterpolationFormatClauseAsync`

**Comment by sharwell** [comment]
I made the following changes:

* Rebased the pull request on the current master branch to correct merge conflicts
* Tweaked wording and code formatting to match current practices
* Added a new test case in 6173a137f8f8700ac1875cfd0e20bde11ab4c19d and simplified surrounding test code a bit
